File: mcp/src/mnemonic_mcp/search_pagination.py
"""Validate explicit search continuation, including byte-limited pages."""

from typing import Any, Protocol, cast

from pydantic import BaseModel, Field, StrictBool, StrictInt


class SearchPagination(BaseModel):
    next_offset: StrictInt | None = Field(default=None, ge=0)
    page_truncated: StrictBool = False


class _Page(Protocol):
    items: list[Any]
    total: int
    limit: int
    offset: int
    next_offset: int | None
    page_truncated: bool


def pagination_matches(page: BaseModel) -> bool:
    if not {"next_offset", "page_truncated"} <= page.model_fields_set:
        return False
    value = cast(_Page, page)
    count = len(value.items)
    expected = min(value.limit, max(0, value.total - value.offset))
    end = value.offset + count
    return (
        0 <= count <= expected
        and (count == expected if not value.page_truncated else 0 < count < expected)
        and value.next_offset == (end if count and end < value.total else None)
    )

File: frontend/lib/search-pagination.ts
import { exactKeys, finiteInteger } from "./wire-guards.ts";

export const SEARCH_PAGINATION_FIELDS = ["next_offset", "page_truncated"] as const;
export type SearchPagination = { next_offset: number | null; page_truncated: boolean };

export function searchPageKeys(page: Record<string, unknown>, keys: readonly string[]): boolean {
  return exactKeys(page, [...keys.filter((key) => key !== "semantic" || Object.hasOwn(page, key)), ...SEARCH_PAGINATION_FIELDS]);
}

export function validSearchPagination(page: Record<string, unknown>): boolean {
  if (!Array.isArray(page.items) || !finiteInteger(page.total) || !finiteInteger(page.limit, 1, 100)
    || !finiteInteger(page.offset) || typeof page.page_truncated !== "boolean") return false;
  const count = page.items.length, expected = Math.min(page.limit, Math.max(0, page.total - page.offset));
  const end = page.offset + count;
  return (page.page_truncated ? count > 0 && count < expected : count === expected)
    && page.next_offset === (count && end < page.total ? end : null);
}

export function searchPagination(page: Record<string, unknown>): SearchPagination {
  if (!validSearchPagination(page)) throw new Error("Mnemonic returned invalid search pagination.");
  return { next_offset: page.next_offset as number | null, page_truncated: page.page_truncated as boolean };
}
